src/typechecker/ct_eval.cpp

```diff
@@ -269,27 +269,77 @@ static TypeFunctionId get_type_function_id(AST::Expr* ast) {
 	}
 }
 
-static void stub_type_function_id(AST::Expr* ast, TypeChecker& tc) {
+static void complete_type_function(AST::Expr* ast, TypeChecker& tc) {
 	switch (ast->type()) {
-
 	case ExprTag::StructExpression: {
 		auto struct_expression = static_cast<AST::StructExpression*>(ast);
-		struct_expression->m_value = tc.core().new_type_function_var();
+		TypeFunctionId tf = struct_expression->m_value;
+
+		auto const& names = struct_expression->m_fields;
+		auto types = compute_monos(struct_expression->m_types, tc);
+
+		int n = names.size();
+		for (int i = 0; i < n; ++i) {
+			tc.core().ll_unify(types[i], tc.core().get_type_function_data(tf).structure[names[i]]);
```

Comment:
Real chunky boi

Reply:
Oof yeah, I just wanted to get it over with

I'll try to split it up a bit.

Reply:
I extracted some variables for now as a stopgap. Maybe a better API can be added in the future.